src/modules/src/error_estimator_ukf.c

```diff
@@ -0,0 +1,1623 @@
+
+  matTp[2 * 3 + 0] = mat[0 * 3 + 2];
+  matTp[2 * 3 + 1] = mat[1 * 3 + 2];
+  matTp[2 * 3 + 2] = mat[2 * 3 + 2];
+}
+
+// compute direction cosine matrix to transfer between lab and body frames
+static void directionCosineMatrix(float *quat, float *dcm)
+{
+  float q0_2 = quat[0] * quat[0];
+  float q1_2 = quat[1] * quat[1];
+  float q2_2 = quat[2] * quat[2];
+  float q3_2 = quat[3] * quat[3];
+
+  // compute elements of the direction cosine matrix, ned to body
+  dcm[0 * 3 + 0] = q0_2 + q1_2 - q2_2 - q3_2;
+  dcm[0 * 3 + 1] = 2 * (quat[1] * quat[2] + quat[0] * quat[3]);
+  dcm[0 * 3 + 2] = 2 * (quat[1] * quat[3] - quat[0] * quat[2]);
+
+  dcm[1 * 3 + 0] = 2 * (quat[1] * quat[2] - quat[0] * quat[3]);
+  dcm[1 * 3 + 1] = q0_2 - q1_2 + q2_2 - q3_2;
+  dcm[1 * 3 + 2] = 2 * (quat[2] * quat[3] + quat[0] * quat[1]);
+
+  dcm[2 * 3 + 0] = 2 * (quat[1] * quat[3] + quat[0] * quat[2]);
+  dcm[2 * 3 + 1] = 2 * (quat[2] * quat[3] - quat[0] * quat[1]);
+  dcm[2 * 3 + 2] = q0_2 - q1_2 - q2_2 + q3_2;
+}
+
+// transform quaternion into euler angles
+static void quatToEuler(float *quat, float *eulerAngles)
+{
+  float q0_2 = quat[0] * quat[0];
+  float q1_2 = quat[1] * quat[1];
+  float q2_2 = quat[2] * quat[2];
+  float q3_2 = quat[3] * quat[3];
+
+  // compute elements of the direction cosine matrix
+  float ele11 = q0_2 + q1_2 - q2_2 - q3_2;
+  float ele21 = 2 * (quat[1] * quat[2] + quat[0] * quat[3]);
+  float ele31 = 2 * (quat[1] * quat[3] - quat[0] * quat[2]);
+  float ele32 = 2 * (quat[2] * quat[3] + quat[0] * quat[1]);
+  float ele33 = q0_2 - q1_2 - q2_2 + q3_2;
+
+  //compute eulerAngles
+  if ((ele32 == 0) & (ele33 == 0))
+  {
+    eulerAngles[0] = 0;
+  }
+  else
+  {
+    eulerAngles[0] = atan2f(ele32, ele33);
+  }
+  eulerAngles[1] = asinf(-ele31);
+
+  if ((ele21 == 0) & (ele11 == 0))
+  {
+    eulerAngles[2] = 0;
+  }
+  else
+  {
+    eulerAngles[2] = atan2f(ele21, ele11);
+  }
+}
+
+// compute quaternion product
+static void multQuat(float *q1, float *q2, float *quatRes)
+{
+  quatRes[0] = q1[0] * q2[0] - q1[1] * q2[1] - q1[2] * q2[2] - q1[3] * q2[3];
+  quatRes[1] = q1[1] * q2[0] + q1[0] * q2[1] - q1[3] * q2[2] + q1[2] * q2[3];
+  quatRes[2] = q1[2] * q2[0] + q1[3] * q2[1] + q1[0] * q2[2] - q1[1] * q2[3];
+  quatRes[3] = q1[3] * q2[0] - q1[2] * q2[1] + q1[1] * q2[2] + q1[0] * q2[3];
+}
+
+// compute quaternion from attitude error
+static void quatFromAtt(float *attVec, float *quat)
+{
+  float absError = sqrtf(attVec[0] * attVec[0] + attVec[1] * attVec[1] + attVec[2] * attVec[2]);
+  float absError2 = absError * absError;
+  float absError3 = absError2 * absError;
+  float absError4 = absError3 * absError;
+
+  float scale = 0.5f - 0.02083333f * absError2 + 0.0002604166f * absError4;
+  quat[0] = 1.0f - 0.125f * absError2 + 0.002604166f * absError4;
+
+  quat[1] = scale * attVec[0];
+  quat[2] = scale * attVec[1];
+  quat[3] = scale * attVec[2];
+}
+
+// Temporary development groups
+LOG_GROUP_START(kalman_states)
```

Comment:
Or you could put it in between #ifdef CONFIG_DEBUG

Reply:
Hey, my suggestion would be to rename the log-group into nav_ukf_states or ukf_states. I prefer the first one since the implementation combines strapdown navigation with a filter so there are two sets of states, the absolute one and the error state.  Would this be ok for you?

Reply:
yes that is possible. I've also put it between the config_debug mode as well